// @dypai-ai/flow/flows/create-booking.flow.ts
import { flow } from "../src/builder.ts";
import { db, email } from "../src/nodes.ts";
import { ref } from "../src/ref.ts";
import { s } from "../src/schema.ts";

export default flow("create-booking")
  .sourceFile("dypai/flows/create-booking.flow.ts")
  .http({ method: "POST", auth: "jwt" })
  .input({
    name: s.string(),
    email: s.email(),
    date: s.date(),
  })
  .output({
    booking: s.object({
      id: s.string(),
      name: s.string(),
      email: s.email(),
      date: s.date(),
    }),
  })
  .step("booking", db.insert({
    table: "public.bookings",
    values: {
      name: ref.input("name"),
      email: ref.input("email"),
      date: ref.input("date"),
      user_id: ref.currentUserId(),
    },
    returning: ["id", "name", "email", "date"],
  }))
  .step("confirmation", email.send({
    to: ref.input("email"),
    subject: "Booking confirmed",
    body: "<p>Hi ${input.name}, your booking for ${input.date} is confirmed.</p>",
  }))
  .return({
    booking: ref.step("booking"),
  });
