// @dypai-ai/flow/src/automation.ts
import { FlowBuilder } from "./builder";
import type { CodeCtx, CodeOptions } from "./code";
import type { Condition } from "./condition";
import type { ScheduleSpec } from "./schedule";
import type { InferSchemaShape, SchemaField } from "./schema";
import type { ComparableRef } from "./ref";
import type { FlowDefinition, FlowStepCall, RefIR, TriggerIR } from "./types";

export type AutomationTrigger =
  | { type: "schedule"; cron: string; timezone?: string }
  | { type: "webhook"; provider?: string; event?: string; name?: string; description?: string }
  | { type: "manual" };

export type SecretRequirementInput = {
  label?: string;
  description?: string;
  sensitive?: boolean;
  required?: boolean;
  metadata?: Record<string, unknown>;
};

export type FileRequirementInput = {
  label?: string;
  description?: string;
  accept?: string;
  required?: boolean;
  metadata?: Record<string, unknown>;
};

export type AutomationRequirement =
  | {
      type: "secret";
      key: string;
      label?: string;
      description?: string;
      sensitive?: boolean;
      required?: boolean;
      metadata?: Record<string, unknown>;
    }
  | {
      type: "file";
      key: string;
      label?: string;
      description?: string;
      accept?: string;
      required?: boolean;
      metadata?: Record<string, unknown>;
    };

export type AutomationNotificationRule = {
  when: string;
  channels: Array<"email" | "in_app" | "agent">;
  condition?: Record<string, unknown>;
  enabled?: boolean;
  metadata?: Record<string, unknown>;
};

export type AutomationHistoryConfig = {
  table?: string;
  idempotencyKey?: string;
  metadata?: Record<string, unknown>;
};

export type AutomationDefinition<TInput = Record<string, unknown>> = {
  kind: "automation";
  name: string;
  title?: string;
  description?: string;
  source_file?: string;
  trigger?: AutomationTrigger;
  requirements: AutomationRequirement[];
  notifications: AutomationNotificationRule[];
  history?: AutomationHistoryConfig;
  flow: FlowDefinition;
};

type WebhookOptions = {
  provider?: string;
  event?: string;
  name?: string;
  description?: string;
};

const ALLOWED_NOTIFICATION_CHANNELS = new Set(["email", "in_app", "agent"]);

function normalizeNotificationChannels(
  channels: string[],
): Array<"email" | "in_app" | "agent"> {
  const normalized: Array<"email" | "in_app" | "agent"> = [];
  for (const channel of channels) {
    const value = String(channel || "").trim().toLowerCase();
    if (
      ALLOWED_NOTIFICATION_CHANNELS.has(value)
      && (value === "email" || value === "in_app" || value === "agent")
      && !normalized.includes(value)
    ) {
      normalized.push(value);
    }
  }
  return normalized.length ? normalized : ["email"];
}

function triggerFromFlow(trigger: TriggerIR): AutomationTrigger | undefined {
  if (trigger.type === "schedule") {
    return {
      type: "schedule",
      cron: trigger.cron,
      ...(trigger.timezone ? { timezone: trigger.timezone } : {}),
    };
  }
  if (trigger.type === "webhook") {
    return {
      type: "webhook",
      ...(trigger.provider ? { provider: trigger.provider } : {}),
      ...(trigger.event ? { event: trigger.event } : {}),
      ...(trigger.name ? { name: trigger.name } : {}),
      ...(trigger.description ? { description: trigger.description } : {}),
    };
  }
  return undefined;
}

function normalizeSecretRequirements(
  record: Record<string, SecretRequirementInput>,
): AutomationRequirement[] {
  return Object.entries(record).map(([key, value]) => ({
    type: "secret" as const,
    key,
    label: value.label,
    description: value.description,
    sensitive: value.sensitive ?? false,
    required: value.required ?? true,
    ...(value.metadata ? { metadata: value.metadata } : {}),
  }));
}

function normalizeFileRequirements(
  record: Record<string, FileRequirementInput>,
): AutomationRequirement[] {
  return Object.entries(record).map(([key, value]) => ({
    type: "file" as const,
    key,
    label: value.label,
    description: value.description,
    accept: value.accept,
    required: value.required ?? true,
    ...(value.metadata ? { metadata: value.metadata } : {}),
  }));
}

export class AutomationBuilder<TInput = Record<string, unknown>> {
  private readonly flow: FlowBuilder<TInput>;
  private titleValue?: string;
  private descriptionValue?: string;
  private requirements: AutomationRequirement[] = [];
  private notifications: AutomationNotificationRule[] = [];
  private historyValue?: AutomationHistoryConfig;
  private inputDeclared = false;
  private triggerDeclared = false;

  constructor(private readonly automationName: string) {
    if (!/^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/.test(automationName)) {
      throw new Error(`Automation name '${automationName}' must be a kebab-case slug.`);
    }
    this.flow = new FlowBuilder(automationName);
  }

  title(value: string): this {
    this.titleValue = value.trim();
    return this;
  }

  description(value: string): this {
    this.descriptionValue = value.trim();
    this.flow.description(value);
    return this;
  }

  sourceFile(path: string): this {
    this.flow.sourceFile(path);
    return this;
  }

  schedule(spec: ScheduleSpec): this {
    this.flow.schedule(spec);
    this.triggerDeclared = true;
    return this;
  }

  webhook(options: WebhookOptions = {}): this {
    this.flow.webhook(options);
    this.triggerDeclared = true;
    return this;
  }

  requiresSecrets(record: Record<string, SecretRequirementInput>): this {
    this.requirements.push(...normalizeSecretRequirements(record));
    return this;
  }

  requiresFiles(record: Record<string, FileRequirementInput>): this {
    this.requirements.push(...normalizeFileRequirements(record));
    return this;
  }

  history(config: AutomationHistoryConfig): this {
    this.historyValue = config;
    return this;
  }

  notify(rule: AutomationNotificationRule): this {
    this.notifications.push({
      ...rule,
      when: String(rule.when || "").trim(),
      channels: normalizeNotificationChannels(rule.channels ?? []),
      enabled: rule.enabled ?? true,
    });
    return this;
  }

  input<TFields extends Record<string, SchemaField>>(
    fields: TFields,
  ): AutomationBuilder<InferSchemaShape<TFields>> {
    this.flow.input(fields);
    this.inputDeclared = true;
    return this as unknown as AutomationBuilder<InferSchemaShape<TFields>>;
  }

  output(fields: Record<string, SchemaField>): this {
    this.flow.output(fields);
    return this;
  }

  response(cardinality: "single" | "many" | "none" | "array"): this {
    this.flow.response(cardinality);
    return this;
  }

  code<TReturn = unknown>(
    fn: (ctx: CodeCtx<TInput>) => TReturn | Promise<TReturn>,
    options?: CodeOptions,
  ): FlowStepCall {
    return this.flow.code(fn, options);
  }

  step(id: string, stepCall: FlowStepCall): this {
    this.flow.step(id, stepCall);
    return this;
  }

  guard(conditionInput: Condition, fallback: Record<string, unknown>): this {
    this.flow.guard(conditionInput, fallback);
    return this;
  }

  return(returns: Record<string, RefIR | ComparableRef>): AutomationDefinition<TInput> {
    if (!this.triggerDeclared) {
      throw new Error(
        `Automation '${this.automationName}' must declare .schedule(...) or .webhook(...) before .return(...).`,
      );
    }
    if (!this.inputDeclared) {
      this.flow.input({});
    }

    const flowDefinition = this.flow.return(returns);
    const trigger = triggerFromFlow(flowDefinition.trigger);
    if (!trigger || (trigger.type !== "schedule" && trigger.type !== "webhook")) {
      throw new Error(
        `Automation '${this.automationName}' must use a schedule or webhook trigger.`,
      );
    }

    return {
      kind: "automation",
      name: this.automationName,
      title: this.titleValue,
      description: this.descriptionValue ?? flowDefinition.endpoint?.description,
      source_file: flowDefinition.source?.file,
      trigger,
      requirements: this.requirements,
      notifications: this.notifications,
      history: this.historyValue,
      flow: flowDefinition,
    };
  }
}

export function automation(name: string): AutomationBuilder {
  return new AutomationBuilder(name);
}

export function isAutomationDefinition(value: unknown): value is AutomationDefinition {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const candidate = value as AutomationDefinition;
  return candidate.kind === "automation"
    && typeof candidate.name === "string"
    && Array.isArray(candidate.requirements)
    && Array.isArray(candidate.notifications)
    && candidate.flow !== undefined
    && typeof candidate.flow.name === "string";
}
