// @dypai-ai/flow/src/branch.ts
import type { Condition } from "./condition";
import type { FlowStepCall, FlowStepDefinition, RefIR } from "./types";
import type { FlowBuilder } from "./builder";
import { toRefIR, type ComparableRef } from "./ref";

export type BranchSnapshot = {
  handle: string;
  steps: FlowStepDefinition[];
  returns?: Record<string, RefIR>;
  returnStepId?: string;
};

export class BranchBuilder {
  readonly steps: FlowStepDefinition[] = [];
  private returnsValue?: Record<string, RefIR>;
  private readonly sourcePath?: string;

  constructor(
    private readonly parent: FlowBuilder<any>,
    readonly handle: string,
    sourcePath?: string,
  ) {
    this.sourcePath = sourcePath;
  }

  step(id: string, stepCall: FlowStepCall): this {
    this.steps.push({
      id,
      call: stepCall.call,
      config: stepCall.config,
      ...(this.sourcePath ? {
        source: {
          file: this.sourcePath,
          stepId: id,
        },
      } : {}),
    });
    return this;
  }

  return(returns: Record<string, RefIR | ComparableRef>): void {
    this.returnsValue = Object.fromEntries(
      Object.entries(returns).map(([key, value]) => [key, toRefIR(value)]),
    );
    const returnStepId = `${this.parent.nextBranchStepId("return")}`;
    this.returnStepId = returnStepId;
    this.steps.push({
      id: returnStepId,
      call: "logic",
      config: {
        operation: "return_response",
        response: this.returnsValue,
      },
      ...(this.sourcePath ? {
        source: {
          file: this.sourcePath,
          stepId: returnStepId,
        },
      } : {}),
    });
  }

  get returns(): Record<string, RefIR> | undefined {
    return this.returnsValue;
  }

  returnStepId?: string;

  snapshot(): BranchSnapshot {
    return {
      handle: this.handle,
      steps: [...this.steps],
      returns: this.returnsValue,
      returnStepId: this.returnStepId,
    };
  }
}

export class WhenBuilder {
  private thenBranch?: BranchSnapshot;
  private elseBranch?: BranchSnapshot;

  constructor(
    private readonly parent: FlowBuilder<any>,
    private readonly logicId: string,
    readonly condition: Condition,
  ) {}

  then(run: (branch: BranchBuilder) => void): this {
    const branch = new BranchBuilder(this.parent, "true", this.parent.getSourcePath());
    run(branch);
    this.thenBranch = branch.snapshot();
    return this;
  }

  else(run: (branch: BranchBuilder) => void): this {
    const branch = new BranchBuilder(this.parent, "false", this.parent.getSourcePath());
    run(branch);
    this.elseBranch = branch.snapshot();
    return this;
  }

  end(): FlowBuilder {
    if (!this.thenBranch?.returns) {
      throw new Error(`Flow '${this.parent.getFlowName()}' when() branch must end with .return(...).`);
    }
    if (!this.elseBranch?.returns) {
      throw new Error(`Flow '${this.parent.getFlowName()}' else() branch must end with .return(...).`);
    }
    this.parent.registerWhenBranch(this.logicId, this.condition, this.thenBranch, this.elseBranch);
    return this.parent;
  }
}
