// @dypai-ai/flow/src/builder.ts
import { BranchBuilder, WhenBuilder, type BranchSnapshot } from "./branch";
import { code as buildCodeStep, type CodeCtx, type CodeOptions } from "./code";
import { conditionToExpression, type Condition } from "./condition";
import { inferOutputSchemaFromReturns, objectSchemaFromFields, type InferSchemaShape, type SchemaField } from "./schema";
import type { ScheduleSpec } from "./schedule";
import type {
  AuthIR,
  FlowBranchGroup,
  FlowDefinition,
  FlowEdgeDefinition,
  FlowStepCall,
  JsonSchema,
  RefIR,
  ResponseIR,
  TriggerIR,
} from "./types";
import { toCondition, toRefIR } from "./ref";

type HttpOptions = {
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE";
  auth: "public" | "jwt";
  roles?: string[];
};

type WebhookOptions = {
  provider?: string;
  event?: string;
  name?: string;
  description?: string;
};

type FlowSegment =
  | { kind: "step"; id: string }
  | { kind: "guard"; logicId: string; fallbackId: string; condition: Condition }
  | { kind: "when"; logicId: string; condition: Condition; thenBranch: BranchSnapshot; elseBranch: BranchSnapshot }
  | {
      kind: "match";
      logicId: string;
      value: RefIR;
      cases: BranchSnapshot[];
      defaultBranch: BranchSnapshot;
    };

export class FlowBuilder<TInput = Record<string, unknown>> {
  private triggerValue?: TriggerIR;
  private authValue?: AuthIR;
  private inputValue?: JsonSchema;
  private outputValue?: JsonSchema;
  private responseValue: ResponseIR = { cardinality: "single" };
  private stepsValue: FlowDefinition["steps"] = [];
  private sourcePath?: string;
  private endpointMeta: FlowDefinition["endpoint"];
  private segments: FlowSegment[] = [];
  private branchCounter = 0;

  constructor(private readonly flowName: string) {}

  getFlowName(): string {
    return this.flowName;
  }

  getSourcePath(): string | undefined {
    return this.sourcePath;
  }

  nextBranchStepId(prefix: string): string {
    this.branchCounter += 1;
    return `__${prefix}_${this.branchCounter}`;
  }

  sourceFile(path: string): this {
    this.sourcePath = path;
    return this;
  }

  http(options: HttpOptions): this {
    this.triggerValue = { type: "http", method: options.method };
    this.authValue = options.auth === "public"
      ? { mode: "public" }
      : {
          mode: "jwt",
          allowedRoles: options.roles !== undefined ? options.roles : [],
        };
    return this;
  }

  schedule(options: ScheduleSpec): this {
    const cron = String(options.cron || "").trim();
    if (!cron) {
      throw new Error(`Flow '${this.flowName}' schedule trigger requires a cron expression.`);
    }
    this.triggerValue = {
      type: "schedule",
      cron,
      ...(options.timezone ? { timezone: options.timezone } : {}),
    };
    this.authValue = { mode: "api_key" };
    return this;
  }

  webhook(options: WebhookOptions = {}): this {
    this.triggerValue = {
      type: "webhook",
      ...(options.provider ? { provider: options.provider } : {}),
      ...(options.event ? { event: options.event } : {}),
      ...(options.name ? { name: options.name } : {}),
      ...(options.description ? { description: options.description } : {}),
    };
    this.authValue = { mode: "api_key" };
    return this;
  }

  description(text: string): this {
    this.endpointMeta = { ...(this.endpointMeta || {}), description: text };
    return this;
  }

  tool(options?: { description?: string }): this {
    this.endpointMeta = {
      ...(this.endpointMeta || {}),
      isTool: true,
      ...(options?.description ? { toolDescription: options.description } : {}),
    };
    return this;
  }

  input<TFields extends Record<string, SchemaField>>(
    fields: TFields,
  ): FlowBuilder<InferSchemaShape<TFields>> {
    this.inputValue = objectSchemaFromFields(fields);
    return this as unknown as FlowBuilder<InferSchemaShape<TFields>>;
  }

  output(fields: Record<string, SchemaField>): this {
    this.outputValue = objectSchemaFromFields(fields);
    return this;
  }

  response(cardinality: ResponseIR["cardinality"] | "array"): this {
    this.responseValue = { cardinality: cardinality === "array" ? "many" : cardinality };
    return this;
  }

  responseCardinality(cardinality: ResponseIR["cardinality"] | "array"): this {
    return this.response(cardinality);
  }

  /** Typed javascript_code step — ctx.input is inferred from .input() on this flow. */
  code<TReturn = unknown>(
    fn: (ctx: CodeCtx<TInput>) => TReturn | Promise<TReturn>,
    options?: CodeOptions,
  ): FlowStepCall {
    return buildCodeStep<TInput, TReturn>(fn, options);
  }

  step(id: string, stepCall: FlowStepCall): this {
    this.stepsValue.push({
      id,
      call: stepCall.call,
      config: stepCall.config,
      ...(this.sourcePath ? {
        source: {
          file: this.sourcePath,
          stepId: id,
        },
      } : {}),
    });
    this.segments.push({ kind: "step", id });
    return this;
  }

  guard(conditionInput: Condition, fallback: Record<string, unknown>): this {
    const condition = toCondition(conditionInput);
    const logicId = this.nextBranchStepId("guard");
    const fallbackId = `${logicId}_fallback`;
    this.stepsValue.push({
      id: logicId,
      call: "logic",
      config: {
        operation: "branch",
        mode: "if_else",
        condition: conditionToExpression(condition),
      },
      ...(this.sourcePath ? { source: { file: this.sourcePath, stepId: logicId } } : {}),
    });
    this.stepsValue.push({
      id: fallbackId,
      call: "logic",
      config: {
        operation: "return_response",
        response: fallback,
      },
      ...(this.sourcePath ? { source: { file: this.sourcePath, stepId: fallbackId } } : {}),
    });
    this.segments.push({ kind: "guard", logicId, fallbackId, condition });
    return this;
  }

  when(conditionInput: Condition): WhenBuilder {
    const condition = toCondition(conditionInput);
    const logicId = this.nextBranchStepId("when");
    this.stepsValue.push({
      id: logicId,
      call: "logic",
      config: {
        operation: "branch",
        mode: "if_else",
        condition: conditionToExpression(condition),
      },
      ...(this.sourcePath ? { source: { file: this.sourcePath, stepId: logicId } } : {}),
    });
    this.segments.push({
      kind: "when",
      logicId,
      condition,
      thenBranch: { handle: "true", steps: [] },
      elseBranch: { handle: "false", steps: [] },
    });
    return new WhenBuilder(this, logicId, condition);
  }

  registerWhenBranch(
    logicId: string,
    condition: Condition,
    thenBranch: BranchSnapshot,
    elseBranch: BranchSnapshot,
  ): void {
    this.stepsValue.push(...thenBranch.steps, ...elseBranch.steps);
    const index = this.segments.findIndex(
      (segment) => segment.kind === "when" && segment.logicId === logicId,
    );
    if (index >= 0) {
      this.segments[index] = {
        kind: "when",
        logicId,
        condition,
        thenBranch,
        elseBranch,
      };
    }
  }

  match<T extends string>(
    value: RefIR,
    cases: Record<T, (branch: BranchBuilder) => void> & { default: (branch: BranchBuilder) => void },
  ): this {
    if (!("default" in cases) || typeof cases.default !== "function") {
      throw new Error(`Flow '${this.flowName}' .match() requires a default branch.`);
    }

    const logicId = this.nextBranchStepId("match");
    const caseEntries = Object.entries(cases).filter(([key]) => key !== "default") as Array<[string, (branch: BranchBuilder) => void]>;
    const branchSnapshots: BranchSnapshot[] = [];

    for (let index = 0; index < caseEntries.length; index += 1) {
      const [caseValue, run] = caseEntries[index];
      const branch = new BranchBuilder(this, `case_${index}`, this.sourcePath);
      run(branch);
      if (!branch.returns) {
        throw new Error(`Flow '${this.flowName}' match branch '${caseValue}' must end with .return(...).`);
      }
      branchSnapshots.push(branch.snapshot());
    }

    const defaultBranchBuilder = new BranchBuilder(this, "default", this.sourcePath);
    cases.default(defaultBranchBuilder);
    if (!defaultBranchBuilder.returns) {
      throw new Error(`Flow '${this.flowName}' match default branch must end with .return(...).`);
    }
    const defaultBranch = defaultBranchBuilder.snapshot();

    this.stepsValue.push({
      id: logicId,
      call: "logic",
      config: {
        operation: "branch",
        mode: "switch",
        switch_on: toRefIR(value),
        cases: caseEntries.map(([caseValue]) => ({ value: caseValue })),
      },
      ...(this.sourcePath ? { source: { file: this.sourcePath, stepId: logicId } } : {}),
    });
    this.stepsValue.push(
      ...branchSnapshots.flatMap((branch) => branch.steps),
      ...defaultBranch.steps,
    );

    this.segments.push({
      kind: "match",
      logicId,
      value: toRefIR(value),
      cases: branchSnapshots,
      defaultBranch,
    });
    return this;
  }

  return(returns: Record<string, RefIR | import("./ref").ComparableRef>): FlowDefinition {
    return this.build(returns);
  }

  build(returns?: Record<string, RefIR | import("./ref").ComparableRef>): FlowDefinition {
    if (!this.triggerValue || !this.authValue) {
      throw new Error(`Flow '${this.flowName}' must declare .http(...), .schedule(...), or .webhook(...) before building.`);
    }
    if (!this.inputValue) {
      throw new Error(`Flow '${this.flowName}' must declare .input(...) before building.`);
    }
    if (!returns || !Object.keys(returns).length) {
      throw new Error(`Flow '${this.flowName}' must declare .return({...}).`);
    }
    const normalizedReturns = Object.fromEntries(
      Object.entries(returns).map(([key, value]) => [key, toRefIR(value)]),
    );
    if (!this.stepsValue.length) {
      throw new Error(`Flow '${this.flowName}' must declare at least one .step(...).`);
    }

    const outputSchema = this.outputValue ?? inferOutputSchemaFromReturns(normalizedReturns);
    const branchGroups = this.buildBranchGroups();
    const edges = this.buildEdges(branchGroups);

    return {
      name: this.flowName,
      ...(this.sourcePath ? {
        source: {
          kind: "flow-ts",
          file: this.sourcePath,
        },
      } : {}),
      ...(this.endpointMeta ? { endpoint: this.endpointMeta } : {}),
      trigger: this.triggerValue,
      auth: this.authValue,
      input: this.inputValue,
      output: outputSchema,
      response: this.responseValue,
      steps: this.stepsValue,
      returns: normalizedReturns,
      edges,
      ...(branchGroups.length ? { branchGroups } : {}),
    };
  }

  private buildBranchGroups(): FlowBranchGroup[] {
    const groups: FlowBranchGroup[] = [];
    for (const segment of this.segments) {
      if (segment.kind === "guard") {
        groups.push({
          id: segment.logicId,
          kind: "guard",
          guardResponse: {},
          branches: [
            { handle: "true", stepIds: [] },
            { handle: "false", stepIds: [segment.fallbackId] },
          ],
        });
      } else if (segment.kind === "when") {
        groups.push({
          id: segment.logicId,
          kind: "when",
          branches: [
            {
              handle: "true",
              stepIds: segment.thenBranch.steps.map((step) => step.id),
              returns: segment.thenBranch.returns,
            },
            {
              handle: "false",
              stepIds: segment.elseBranch.steps.map((step) => step.id),
              returns: segment.elseBranch.returns,
            },
          ],
        });
      } else if (segment.kind === "match") {
        groups.push({
          id: segment.logicId,
          kind: "match",
          branches: [
            ...segment.cases.map((branch) => ({
              handle: branch.handle,
              stepIds: branch.steps.map((step) => step.id),
              returns: branch.returns,
            })),
            {
              handle: "default",
              stepIds: segment.defaultBranch.steps.map((step) => step.id),
              returns: segment.defaultBranch.returns,
            },
          ],
        });
      }
    }
    return groups;
  }

  private buildEdges(branchGroups: FlowBranchGroup[]): FlowEdgeDefinition[] {
    if (!branchGroups.length) {
      const edges: FlowEdgeDefinition[] = [];
      if (this.stepsValue.length) {
        edges.push({ from: "$trigger", to: this.stepsValue[0].id });
        for (let i = 0; i < this.stepsValue.length - 1; i += 1) {
          edges.push({ from: this.stepsValue[i].id, to: this.stepsValue[i + 1].id });
        }
        edges.push({ from: this.stepsValue[this.stepsValue.length - 1].id, to: "return" });
      }
      return edges;
    }

    const edges: FlowEdgeDefinition[] = [];
    const branchOwnedStepIds = new Set<string>();
    for (const group of branchGroups) {
      for (const branch of group.branches) {
        for (const stepId of branch.stepIds) branchOwnedStepIds.add(stepId);
      }
    }

    const connectChain = (fromId: string, stepIds: string[]): string => {
      let current = fromId;
      for (const stepId of stepIds) {
        edges.push({ from: current, to: stepId });
        current = stepId;
      }
      return current;
    };

    const wireBranch = (
      logicId: string,
      branch: BranchSnapshot,
      handle: FlowEdgeDefinition["when"] | FlowEdgeDefinition["case"],
      handleKey: "when" | "case",
    ) => {
      const ids = branch.steps.map((step) => step.id);
      if (!ids.length) return;
      edges.push({
        from: logicId,
        to: ids[0],
        ...(handleKey === "when" ? { when: handle as "true" | "false" } : { case: handle as string }),
      });
      connectChain(ids[0], ids.slice(1));
    };

    let incoming = "$trigger";
    let segmentIndex = 0;
    while (segmentIndex < this.segments.length) {
      const segment = this.segments[segmentIndex];

      if (segment.kind === "step" && !branchOwnedStepIds.has(segment.id)) {
        edges.push({ from: incoming, to: segment.id });
        incoming = segment.id;
        segmentIndex += 1;
        continue;
      }

      if (segment.kind === "guard") {
        edges.push({ from: incoming, to: segment.logicId });
        edges.push({ from: segment.logicId, to: segment.fallbackId, when: "false" });

        const truePath: string[] = [];
        let lookahead = segmentIndex + 1;
        while (lookahead < this.segments.length) {
          const next = this.segments[lookahead];
          if (next.kind !== "step" || branchOwnedStepIds.has(next.id)) break;
          truePath.push(next.id);
          lookahead += 1;
        }

        if (truePath.length) {
          edges.push({ from: segment.logicId, to: truePath[0], when: "true" });
          incoming = connectChain(truePath[0], truePath.slice(1));
          segmentIndex = lookahead;
        } else {
          edges.push({ from: segment.logicId, to: "return", when: "true" });
          incoming = "return";
          segmentIndex += 1;
        }
        continue;
      }

      if (segment.kind === "when") {
        edges.push({ from: incoming, to: segment.logicId });
        wireBranch(segment.logicId, segment.thenBranch, "true", "when");
        wireBranch(segment.logicId, segment.elseBranch, "false", "when");
        incoming = segment.logicId;
        segmentIndex += 1;
        continue;
      }

      if (segment.kind === "match") {
        edges.push({ from: incoming, to: segment.logicId });
        for (const branch of segment.cases) {
          wireBranch(segment.logicId, branch, branch.handle, "case");
        }
        wireBranch(segment.logicId, segment.defaultBranch, "default", "case");
        incoming = segment.logicId;
        segmentIndex += 1;
        continue;
      }

      segmentIndex += 1;
    }

    const terminalReturnResponse = this.stepsValue.some(
      (step) => step.config.operation === "return_response",
    );
    const hasTerminalBranch = this.segments.some(
      (segment) => segment.kind === "when" || segment.kind === "match",
    );

    if (incoming !== "return" && !terminalReturnResponse && !hasTerminalBranch) {
      edges.push({ from: incoming, to: "return" });
    } else if (
      incoming !== "return"
      && !hasTerminalBranch
      && !this.stepsValue.some((step) => step.id === incoming && step.config.operation === "return_response")
    ) {
      edges.push({ from: incoming, to: "return" });
    }

    return edges;
  }
}

export function flow(name: string): FlowBuilder {
  if (!/^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/.test(name)) {
    throw new Error(`Flow name '${name}' must be a kebab-case slug.`);
  }
  return new FlowBuilder(name);
}
