// @dypai-ai/flow/src/code.ts
import type { FlowStepCall } from "./types";

/** Workflow context available inside typed code() steps. */
export type CodeCtx<TInput = Record<string, unknown>> = {
  /** Original HTTP request — always the endpoint input, not the previous node output. */
  input: TInput;
  /** Automation setup injected for scheduled/webhook automations, when present. */
  automation?: {
    files?: Record<string, {
      bucket?: string;
      storage_path?: string;
      object_id?: string;
      original_filename?: string;
      content_type?: string;
      size_bytes?: number;
      file_sha256?: string;
    }>;
    secrets?: Record<string, { secret_name?: string }>;
    requirements?: Record<string, unknown>;
    [key: string]: unknown;
  };
  /** Outputs from previously executed nodes, keyed by step id. */
  nodes: Record<string, unknown>;
  user: {
    id: string;
    email?: string;
    role?: string;
    metadata?: Record<string, unknown>;
  } | null;
  env: Record<string, string>;
  db?: Record<string, (...args: unknown[]) => unknown>;
  http?: Record<string, (...args: unknown[]) => unknown>;
  dypai?: Record<string, (...args: unknown[]) => unknown>;
  storage?: Record<string, (...args: unknown[]) => unknown>;
  /** ISO timestamp captured when the node starts executing. */
  now: string;
};

export type CodeOptions = {
  timeoutMs?: number;
  env?: Record<string, string>;
};

const FLOW_SCOPE_LEAK = /(^|[^A-Za-z0-9_$.])(ref|db|s|flow|capability|code)\./;

function validateCodeFunctionSource(source: string): void {
  if (FLOW_SCOPE_LEAK.test(source)) {
    throw new Error(
      "code() functions cannot reference flow builder helpers (ref, db, s, etc.). "
      + "Use ctx.input / ctx.nodes inside the function body only.",
    );
  }
  if (/\b(import|require)\b/.test(source)) {
    throw new Error("code() functions cannot use import or require.");
  }
}

function serializeCodeFunction(fn: (...args: unknown[]) => unknown): string {
  const source = fn.toString().trim();
  validateCodeFunctionSource(source);
  return [
    "async function main({ input, automation, nodes, user, env, db, http, dypai, storage, now }) {",
    `  const __flowFn__ = ${source};`,
    "  return await __flowFn__({ input, automation, nodes, user, env, db, http, dypai, storage, now });",
    "}",
  ].join("\n");
}

/**
 * Typed JavaScript code step — compiles to javascript_code with a main(ctx) body.
 * TypeScript validates ctx.input fields at authoring time.
 */
export function code<TInput = Record<string, unknown>, TReturn = unknown>(
  fn: (ctx: CodeCtx<TInput>) => TReturn | Promise<TReturn>,
  options?: CodeOptions,
): FlowStepCall {
  const serialized = serializeCodeFunction(fn as (...args: unknown[]) => unknown);
  const envVars = options?.env
    ? Object.entries(options.env).map(([key, value]) => ({ key, value }))
    : [];

  return {
    call: "javascript_code",
    config: {
      code: serialized,
      input_data: {},
      timeout_ms: options?.timeoutMs ?? 5000,
      runtime: "vm",
      ...(envVars.length ? { env_vars: envVars } : { env_vars: [] }),
    },
  };
}

export type { CodeCtx as CodeContext };
