// @dypai-ai/flow/src/condition.ts
import type { RefIR } from "./types";

export type ConditionOp =
  | "eq"
  | "ne"
  | "gt"
  | "gte"
  | "lt"
  | "lte"
  | "in"
  | "matches"
  | "notEmpty";

export type SimpleCondition = {
  left: RefIR;
  op: ConditionOp;
  right?: RefIR;
};

export type Condition =
  | SimpleCondition
  | { and: Condition[] }
  | { or: Condition[] }
  | { expr: string };

function refToPlaceholder(ref: RefIR): string {
  switch (ref.kind) {
    case "input":
      return `\${input.${ref.path.join(".")}}`;
    case "currentUserId":
      return "${current_user_id}";
    case "step":
      return ref.path?.length
        ? `\${vars.${ref.stepId}.${ref.path.join(".")}}`
        : `\${vars.${ref.stepId}}`;
    case "literal":
      return typeof ref.value === "string" ? JSON.stringify(ref.value) : String(ref.value);
    case "env":
      return `\${env.${ref.name}}`;
    default:
      return "false";
  }
}

function literalToJs(ref: RefIR): string {
  if (ref.kind !== "literal") return refToPlaceholder(ref);
  if (typeof ref.value === "string") return JSON.stringify(ref.value);
  return String(ref.value);
}

function simpleConditionToExpression(condition: SimpleCondition): string {
  const left = refToPlaceholder(condition.left);
  switch (condition.op) {
    case "eq":
      return `${left} == ${literalToJs(condition.right ?? { kind: "literal", value: null })}`;
    case "ne":
      return `${left} != ${literalToJs(condition.right ?? { kind: "literal", value: null })}`;
    case "gt":
      return `${left} > ${literalToJs(condition.right ?? { kind: "literal", value: 0 })}`;
    case "gte":
      return `${left} >= ${literalToJs(condition.right ?? { kind: "literal", value: 0 })}`;
    case "lt":
      return `${left} < ${literalToJs(condition.right ?? { kind: "literal", value: 0 })}`;
    case "lte":
      return `${left} <= ${literalToJs(condition.right ?? { kind: "literal", value: 0 })}`;
    case "in": {
      const values = condition.right?.kind === "literal" && Array.isArray(condition.right.value)
        ? condition.right.value
        : [];
      return `${JSON.stringify(values)}.includes(String(${left}))`;
    }
    case "matches": {
      const pattern = condition.right?.kind === "literal" ? String(condition.right.value) : "";
      return `new RegExp(${JSON.stringify(pattern)}).test(String(${left}))`;
    }
    case "notEmpty":
      return `${left} !== undefined && ${left} !== null && String(${left}).trim() !== ""`;
    default:
      return "false";
  }
}

export function conditionToExpression(condition: Condition): string {
  if ("expr" in condition && typeof condition.expr === "string") return condition.expr;
  if ("and" in condition && Array.isArray(condition.and)) {
    return condition.and.map((item) => `(${conditionToExpression(item)})`).join(" && ");
  }
  if ("or" in condition && Array.isArray(condition.or)) {
    return condition.or.map((item) => `(${conditionToExpression(item)})`).join(" || ");
  }
  return simpleConditionToExpression(condition as SimpleCondition);
}

export function conditionToBuilderRows(condition: Condition): Array<{
  left: string;
  operator: string;
  right?: unknown;
}> {
  if ("expr" in condition) {
    return [{ left: condition.expr, operator: "is_true" }];
  }
  if ("and" in condition && Array.isArray(condition.and)) {
    return condition.and.flatMap((item) => conditionToBuilderRows(item));
  }
  if ("or" in condition && Array.isArray(condition.or)) {
    return condition.or.flatMap((item) => conditionToBuilderRows(item));
  }
  const simple = condition as SimpleCondition;
  const operatorMap: Record<ConditionOp, string> = {
    eq: "eq",
    ne: "neq",
    gt: "gt",
    gte: "gte",
    lt: "lt",
    lte: "lte",
    in: "contains",
    matches: "contains",
    notEmpty: "is_not_empty",
  };
  const row: { left: string; operator: string; right?: unknown } = {
    left: refToPlaceholder(simple.left),
    operator: operatorMap[simple.op],
  };
  if (simple.right && simple.op !== "notEmpty") {
    row.right = simple.right.kind === "literal"
      ? simple.right.value
      : refToPlaceholder(simple.right);
  }
  return [row];
}

export function isCondition(value: unknown): value is Condition {
  if (!value || typeof value !== "object") return false;
  if ("expr" in value && typeof (value as { expr?: unknown }).expr === "string") return true;
  if ("and" in value && Array.isArray((value as { and?: unknown }).and)) return true;
  if ("or" in value && Array.isArray((value as { or?: unknown }).or)) return true;
  return "left" in value && "op" in value;
}

export type ComposableCondition = Condition & {
  and(other: Condition): Condition;
  or(other: Condition): Condition;
};

export function composableCondition(condition: Condition): ComposableCondition {
  return Object.assign({}, condition, {
    and(other: Condition) {
      return { and: [condition, other] };
    },
    or(other: Condition) {
      return { or: [condition, other] };
    },
  });
}
