// @dypai-ai/flow/src/index.ts
export { flow, FlowBuilder } from "./builder";
export {
  automation,
  AutomationBuilder,
  isAutomationDefinition,
  type AutomationDefinition,
  type AutomationHistoryConfig,
  type AutomationNotificationRule,
  type AutomationRequirement,
  type AutomationTrigger,
  type FileRequirementInput,
  type SecretRequirementInput,
} from "./automation";
export { BranchBuilder, WhenBuilder } from "./branch";
export type { Condition, ConditionOp } from "./condition";
export { capability, db, email, storage, dypai, type DbQueryCall } from "./nodes";
export { code, type CodeCtx, type CodeContext, type CodeOptions } from "./code";
export { generateCapabilityWrappersSource } from "./generateCapabilityWrappers";
export { ref, toRefIR, toCondition } from "./ref";
export type { ComparableRef } from "./ref";
export { every, type ScheduleOptions, type ScheduleSpec } from "./schedule";
export { inferOutputSchemaFromReturns, objectSchemaFromFields, s, type InferSchemaShape, type SchemaField } from "./schema";
export type {
  AuthIR,
  CapabilityCatalog,
  FlowBranchGroup,
  FlowDefinition,
  FlowEdgeDefinition,
  FlowStepCall,
  FlowStepDefinition,
  JsonSchema,
  RefIR,
  ResponseIR,
  TriggerIR,
} from "./types";
