// @dypai-ai/flow/src/nodes.ts
export type { FlowStepCall } from "./types";

import type { FlowStepCall } from "./types";

type DbQueryOutputCardinality = "single" | "many" | "zero_or_one";
type DbQueryConfig = {
  sql: string;
  params?: Record<string, unknown>;
  outputCardinality?: DbQueryOutputCardinality;
};

export type DbQueryCall = FlowStepCall & {
  single(): FlowStepCall;
  maybeSingle(): FlowStepCall;
  many(): FlowStepCall;
};

export function capability(
  groupId: string,
  operation: string,
  config: Record<string, unknown>,
): FlowStepCall {
  return {
    call: `${groupId}.${operation}`,
    config,
  };
}

function makeDbQueryCall(config: DbQueryConfig): DbQueryCall {
  const base = capability("dypai.db", "query", config);
  return Object.assign(base, {
    single(): FlowStepCall {
      return makeDbQueryCall({ ...config, outputCardinality: "single" });
    },
    maybeSingle(): FlowStepCall {
      return makeDbQueryCall({ ...config, outputCardinality: "zero_or_one" });
    },
    many(): FlowStepCall {
      return makeDbQueryCall({ ...config, outputCardinality: "many" });
    },
  });
}

function query(config: DbQueryConfig): DbQueryCall {
  return makeDbQueryCall(config);
}

query.single = (config: Omit<DbQueryConfig, "outputCardinality">) => {
  return makeDbQueryCall({ ...config, outputCardinality: "single" });
};

query.maybeSingle = (config: Omit<DbQueryConfig, "outputCardinality">) => {
  return makeDbQueryCall({ ...config, outputCardinality: "zero_or_one" });
};

query.many = (config: Omit<DbQueryConfig, "outputCardinality">) => {
  return makeDbQueryCall({ ...config, outputCardinality: "many" });
};

export const db = {
  query,
  insert(config: { table: string; values: Record<string, unknown>; returning?: string[] }) {
    return capability("dypai.db", "insert", config);
  },
  update(config: {
    table: string;
    where: Record<string, unknown>;
    set: Record<string, unknown>;
    returning?: string[];
  }) {
    return capability("dypai.db", "update", config);
  },
  list(config: {
    table: string;
    where?: Record<string, unknown>;
    orderBy?: string;
    limit?: unknown;
    offset?: unknown;
  }) {
    return capability("dypai.db", "list", config);
  },
};

export const email = {
  send(config: {
    to: unknown;
    subject: string;
    /** HTML body. Supports ${input.*} / ${nodes.*} placeholders, resolved by the engine. */
    body: string;
    cc?: unknown;
    bcc?: unknown;
    reply_to?: unknown;
    /** Display name for the sender. The address itself is managed by DYPAI. */
    from_name?: string;
    /** Array of { filename, content } (content base64) for documents of record (invoices, etc.). */
    attachments?: unknown;
  }) {
    return capability("dypai.email", "send", config);
  },
};

type StoragePathConfig = {
  bucket: string;
  path?: unknown;
  fileId?: unknown;
};

type StorageDownloadConfig = StoragePathConfig & {
  expiresMinutes?: number;
  /** When true (default), signed URL forces download. When false, inline/preview. */
  download?: boolean;
};

function storageDownloadConfig(config: StorageDownloadConfig) {
  return {
    operation: "download",
    bucket: config.bucket,
    file_path: config.path,
    file_id: config.fileId,
    expires_minutes: config.expiresMinutes,
    download: config.download ?? true,
  };
}

export const storage = {
  upload(config: { bucket: string; path?: unknown; prefix?: unknown; entityId?: unknown }) {
    return capability("dypai.storage", "upload", {
      operation: "upload",
      bucket: config.bucket,
      file_path: config.path,
      prefix: config.prefix,
      entity_id: config.entityId,
    });
  },
  download(config: StorageDownloadConfig) {
    return capability("dypai.storage", "download", storageDownloadConfig(config));
  },
  /** Signed read URL without forcing attachment download (inline/preview). */
  signedUrl(config: Omit<StorageDownloadConfig, "download">) {
    return capability("dypai.storage", "download", storageDownloadConfig({ ...config, download: false }));
  },
  delete(config: StoragePathConfig) {
    return capability("dypai.storage", "delete", {
      operation: "delete",
      bucket: config.bucket,
      file_path: config.path,
      file_id: config.fileId,
    });
  },
  read(config: StoragePathConfig) {
    return capability("dypai.storage", "read", {
      operation: "read",
      bucket: config.bucket,
      file_path: config.path,
      file_id: config.fileId,
    });
  },
};

export const dypai = {
  db,
  email,
  storage,
  capability,
};
