// @dypai-ai/flow/src/ref.ts
import type { ComposableCondition, Condition, ConditionOp } from "./condition";
import { composableCondition, isCondition } from "./condition";
import type { RefIR } from "./types";

export type ComparableRef = RefIR & {
  toRef(): RefIR;
  eq(value: unknown): ComposableCondition;
  ne(value: unknown): ComposableCondition;
  gt(value: number): ComposableCondition;
  gte(value: number): ComposableCondition;
  lt(value: number): ComposableCondition;
  lte(value: number): ComposableCondition;
  in(values: unknown[]): ComposableCondition;
  matches(pattern: string): ComposableCondition;
  notEmpty(): ComposableCondition;
};

function compare(ref: RefIR, op: ConditionOp, value: unknown): ComposableCondition {
  return composableCondition({
    left: ref,
    op,
    right: { kind: "literal", value },
  });
}

function wrapRef(ref: RefIR): ComparableRef {
  return {
    ...ref,
    toRef() {
      return ref;
    },
    eq(value: unknown) {
      return compare(ref, "eq", value);
    },
    ne(value: unknown) {
      return compare(ref, "ne", value);
    },
    gt(value: number) {
      return compare(ref, "gt", value);
    },
    gte(value: number) {
      return compare(ref, "gte", value);
    },
    lt(value: number) {
      return compare(ref, "lt", value);
    },
    lte(value: number) {
      return compare(ref, "lte", value);
    },
    in(values: unknown[]) {
      return compare(ref, "in", values);
    },
    matches(pattern: string) {
      return compare(ref, "matches", pattern);
    },
    notEmpty() {
      return composableCondition({ left: ref, op: "notEmpty" });
    },
  };
}

export const ref = {
  input(key: string, ...path: string[]): ComparableRef {
    return wrapRef({ kind: "input", path: [key, ...path] });
  },

  currentUserId(): ComparableRef {
    return wrapRef({ kind: "currentUserId" });
  },

  step(stepId: string, ...path: string[]): ComparableRef {
    return wrapRef(
      path.length
        ? { kind: "step", stepId, path }
        : { kind: "step", stepId },
    );
  },

  literal(value: unknown): ComparableRef {
    return wrapRef({ kind: "literal", value });
  },

  env(name: string): ComparableRef {
    return wrapRef({ kind: "env", name });
  },

  expr(expression: string): Condition {
    return { expr: expression };
  },
};

export type { RefIR };

export function toRefIR(value: RefIR | ComparableRef): RefIR {
  if ("toRef" in value && typeof value.toRef === "function") {
    return value.toRef();
  }
  return value;
}

export function toCondition(value: Condition | ComparableRef): Condition {
  if (isCondition(value)) return value;
  if ("kind" in value) return wrapRef(value as RefIR).notEmpty();
  return value;
}
