// @dypai-ai/flow/src/schema.ts
import type { JsonSchema } from "./types";

export type SchemaField<T = unknown> = JsonSchema & {
  optional?: boolean;
  readonly __type?: T;
};

type InferField<F> = F extends SchemaField<infer U> ? U : unknown;

export type InferSchemaShape<T extends Record<string, SchemaField>> = {
  [K in keyof T as T[K] extends SchemaField & { optional: true } ? never : K]: InferField<T[K]>;
} & {
  [K in keyof T as T[K] extends SchemaField & { optional: true } ? K : never]?: InferField<T[K]>;
};

function isSchemaField(value: unknown): value is SchemaField {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

export function objectSchemaFromFields(fields: Record<string, SchemaField>): JsonSchema {
  const properties: Record<string, JsonSchema> = {};
  const required: string[] = [];

  for (const [key, field] of Object.entries(fields)) {
    const { optional, ...schema } = field;
    properties[key] = schema;
    if (!optional) required.push(key);
  }

  return {
    type: "object",
    additionalProperties: false,
    ...(required.length ? { required } : {}),
    properties,
  };
}

export const s = {
  string(options?: { format?: string; minLength?: number }): SchemaField<string> {
    return {
      type: "string",
      ...(options?.format ? { format: options.format } : {}),
      ...(options?.minLength ? { minLength: options.minLength } : {}),
    };
  },

  email(): SchemaField<string> {
    return { type: "string", format: "email" };
  },

  number(): SchemaField<number> {
    return { type: "number" };
  },

  integer(): SchemaField<number> {
    return { type: "integer" };
  },

  boolean(): SchemaField<boolean> {
    return { type: "boolean" };
  },

  date(): SchemaField<string> {
    return { type: "string", format: "date" };
  },

  optional<T>(field: SchemaField<T>): SchemaField<T | undefined> {
    return { ...field, optional: true };
  },

  array<T>(items: SchemaField<T>): SchemaField<T[]> {
    const { optional, ...itemSchema } = items;
    return {
      type: "array",
      items: itemSchema,
      ...(optional ? { optional: true } : {}),
    };
  },

  object(properties: Record<string, SchemaField>, options?: { required?: string[] }): SchemaField<Record<string, unknown>> {
    const schema = objectSchemaFromFields(properties);
    if (options?.required) {
      schema.required = options.required;
    }
    return schema;
  },

  uuid(): SchemaField<string> {
    return { type: "string", format: "uuid" };
  },
};

export function inferOutputSchemaFromReturns(
  returns: Record<string, unknown>,
): JsonSchema {
  const properties: Record<string, JsonSchema> = {};
  for (const key of Object.keys(returns)) {
    properties[key] = { type: "object" };
  }
  return {
    type: "object",
    additionalProperties: false,
    properties,
  };
}

export function isSchemaFieldValue(value: unknown): value is SchemaField {
  return isSchemaField(value);
}
