// @dypai-ai/flow/src/schedule.ts
export type ScheduleSpec = {
  cron: string;
  timezone?: string;
};

export type ScheduleOptions = {
  timezone?: string;
};

const DAYS: Record<string, number> = {
  sunday: 0,
  sun: 0,
  monday: 1,
  mon: 1,
  tuesday: 2,
  tue: 2,
  wednesday: 3,
  wed: 3,
  thursday: 4,
  thu: 4,
  friday: 5,
  fri: 5,
  saturday: 6,
  sat: 6,
};

function withTimezone(cron: string, options?: ScheduleOptions): ScheduleSpec {
  return {
    cron,
    ...(options?.timezone ? { timezone: options.timezone } : {}),
  };
}

function parseTime(time: string): { hour: number; minute: number } {
  const match = /^(\d{1,2}):(\d{2})$/.exec(time.trim());
  if (!match) {
    throw new Error(`Invalid time '${time}'. Use HH:mm, for example "09:30".`);
  }
  const hour = Number(match[1]);
  const minute = Number(match[2]);
  if (!Number.isInteger(hour) || hour < 0 || hour > 23 || !Number.isInteger(minute) || minute < 0 || minute > 59) {
    throw new Error(`Invalid time '${time}'. Hour must be 0-23 and minute 0-59.`);
  }
  return { hour, minute };
}

function positiveInteger(value: number, label: string): number {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${label} must be a positive integer.`);
  }
  return value;
}

function dayOfWeek(value: string | number): number {
  if (typeof value === "number") {
    if (!Number.isInteger(value) || value < 0 || value > 6) {
      throw new Error("Weekday number must be 0-6, where 0 is Sunday.");
    }
    return value;
  }
  const normalized = value.trim().toLowerCase();
  const mapped = DAYS[normalized];
  if (mapped === undefined) {
    throw new Error(`Invalid weekday '${value}'. Use monday, tuesday, ... or 0-6.`);
  }
  return mapped;
}

function minutes(value: number, options?: ScheduleOptions): ScheduleSpec {
  const minutesValue = positiveInteger(value, "minutes");
  if (minutesValue > 59) throw new Error("minutes must be between 1 and 59.");
  return withTimezone(`*/${minutesValue} * * * *`, options);
}

function hour(options?: ScheduleOptions): ScheduleSpec {
  return withTimezone("0 * * * *", options);
}

function hours(value: number, options?: ScheduleOptions): ScheduleSpec {
  const hoursValue = positiveInteger(value, "hours");
  if (hoursValue > 23) throw new Error("hours must be between 1 and 23.");
  return withTimezone(`0 */${hoursValue} * * *`, options);
}

function dayAt(time: string, options?: ScheduleOptions): ScheduleSpec {
  const { hour: hourValue, minute } = parseTime(time);
  return withTimezone(`${minute} ${hourValue} * * *`, options);
}

function weekOn(day: string | number, time: string, options?: ScheduleOptions): ScheduleSpec {
  const { hour: hourValue, minute } = parseTime(time);
  return withTimezone(`${minute} ${hourValue} * * ${dayOfWeek(day)}`, options);
}

function monthOn(day: number, time: string, options?: ScheduleOptions): ScheduleSpec {
  const monthDay = positiveInteger(day, "day");
  if (monthDay > 31) throw new Error("day must be between 1 and 31.");
  const { hour: hourValue, minute } = parseTime(time);
  return withTimezone(`${minute} ${hourValue} ${monthDay} * *`, options);
}

function cron(cronExpression: string, options?: ScheduleOptions): ScheduleSpec {
  const normalized = cronExpression.trim();
  if (!normalized) throw new Error("cron must not be empty.");
  return withTimezone(normalized, options);
}

export const every = {
  minutes,
  hour,
  hours,
  dayAt,
  dailyAt: dayAt,
  weekOn,
  weeklyOn: weekOn,
  monthOn,
  monthlyOn: monthOn,
  cron,
};
