// @dypai-ai/flow/src/types.ts
export type FlowStepCall = {
  call: string;
  config: Record<string, unknown>;
};

export type JsonSchema = Record<string, unknown>;

export type RefIR =
  | { kind: "input"; path: string[] }
  | { kind: "currentUserId" }
  | { kind: "step"; stepId: string; path?: string[] }
  | { kind: "literal"; value: unknown }
  | { kind: "env"; name: string };

export type AuthIR =
  | { mode: "public" }
  | { mode: "jwt"; allowedRoles: string[] }
  | { mode: "api_key" };

export type TriggerIR =
  | { type: "http"; method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE" }
  | { type: "schedule"; cron: string; timezone?: string }
  | {
      type: "webhook";
      provider?: string;
      event?: string;
      name?: string;
      description?: string;
    };

export type ResponseIR = {
  cardinality?: "single" | "many" | "none";
};

export type FlowStepDefinition = {
  id: string;
  call: string;
  config: Record<string, unknown>;
  source?: {
    file?: string;
    line?: number;
    column?: number;
    yamlPath?: string;
    stepId?: string;
  };
};

export type FlowEdgeDefinition = {
  from: string;
  to: string;
  when?: "true" | "false";
  case?: string;
};

export type FlowBranchGroup = {
  id: string;
  kind: "guard" | "when" | "match";
  guardResponse?: Record<string, RefIR | unknown>;
  branches: Array<{
    handle: string;
    stepIds: string[];
    returns?: Record<string, RefIR>;
  }>;
};

export type FlowDefinition = {
  name: string;
  source?: {
    kind: "flow-ts";
    file: string;
  };
  endpoint?: {
    description?: string;
    isTool?: boolean;
    toolDescription?: string;
  };
  trigger: TriggerIR;
  auth: AuthIR;
  input: JsonSchema;
  output: JsonSchema;
  response?: ResponseIR;
  steps: FlowStepDefinition[];
  returns: Record<string, RefIR>;
  edges?: FlowEdgeDefinition[];
  branchGroups?: FlowBranchGroup[];
};

export type CapabilityCatalog = {
  version?: string;
  content_hash?: string;
  groups: Array<{
    id: string;
    namespace: string;
    resource: string;
    operations: Array<{ operation: string }>;
  }>;
};
